contracts/governance/BaseWeightedMultisig.sol

```diff
@@ -0,0 +1,185 @@
+     * @param hash The signers hash
+     * @return uint256 The epoch for the given signers hash
+     */
+    function signersEpochForHash(bytes32 hash) external view returns (uint256) {
+        return _baseWeightedStorage().signersEpochForHash[hash];
+    }
+
+    /*************************\
+    |* Integration Functions *|
+    \*************************/
+    /*
+     * @notice This function takes messageHash and proof data and reverts if proof is invalid
+     * @param messageHash The hash of the message that was signed
+     * @param weightedSigners The weighted signers data
+     * @param signatures The signatures data
+     * @return isLatestSigners True if provided signers are the current ones
+     */
+    function _validateProof(
+        bytes32 messageHash,
+        bytes calldata weightedSigners,
+        bytes[] calldata signatures
+    ) internal view returns (bool isLatestSigners) {
+        WeightedMultisigStorage storage $ = _baseWeightedStorage();
+        bytes32 signersHash = keccak256(weightedSigners);
+        uint256 signersEpoch = $.signersEpochForHash[signersHash];
+        uint256 currentEpoch = $.currentSignersEpoch;
+
+        isLatestSigners = signersEpoch == currentEpoch;
+
+        if (signatures.length == 0) revert MalformedSignatures();
+        if (signersEpoch == 0 || currentEpoch - signersEpoch >= OLD_SIGNERS_RETENTION || weightedSigners.length == 0) revert InvalidSigners();
+
+        WeightedSigners memory signers = abi.decode(weightedSigners, (WeightedSigners));
+
+        _validateSignatures(messageHash, signers, signatures);
+    }
+
+    /*
+     * @notice This function rotates the current signers with a new set of signers
+     * @param newWeightedSigners The new weighted signers data
+     */
+    function _rotateSigners(bytes memory newWeightedSigners) internal {
+        WeightedMultisigStorage storage $ = _baseWeightedStorage();
+        WeightedSigners memory newSet = abi.decode(newWeightedSigners, (WeightedSigners));
+
+        // signers must be sorted binary or alphabetically in lower case
+        if (newSet.signers.length == 0 || !_isSortedAscAndContainsNoDuplicate(newSet.signers)) revert InvalidSigners();
+
+        uint256 length = newSet.signers.length;
+        uint256 totalWeight;
+        for (uint256 i; i < length; ++i) {
+            uint256 weight = newSet.signers[i].weight;
+
+            if (weight == 0) revert InvalidWeights();
+
+            totalWeight = totalWeight + weight;
+        }
+        if (newSet.threshold == 0 || totalWeight < newSet.threshold) revert InvalidThreshold();
+
+        bytes32 newSignersHash = keccak256(newWeightedSigners);
+
+        uint256 epoch = $.currentSignersEpoch + 1;
+        // slither-disable-next-line costly-loop
+        $.currentSignersEpoch = epoch;
+        $.hashForSignersEpoch[epoch] = newSignersHash;
+        $.signersEpochForHash[newSignersHash] = epoch;
+
+        emit SignersRotated(newSet);
```

Comment:
emit earlier and include the old set as well?

Reply:
There is no old set, as we only have a commit for it. And I think we don't need to emit earlier as there is no function calls made prior to the emit.